presto-spark/src/main/java/com/facebook/presto/spark/planner/SparkRddPlanner.java

```diff
@@ -0,0 +1,254 @@
+        RddFactory rddFactory = new RddFactory(
+                session,
+                sparkTaskRequestJsonCodec,
+                sparkContext,
+                taskCompilerFactory,
+                getHashPartitionCount(session),
+                taskStatsCollector,
+                preparedPlan.getTableWriteInfo());
+        return rddFactory.createRdd(preparedPlan.getPlan());
+    }
+
+    private static class RddFactory
+    {
+        private final Session session;
+        private final JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec;
+        private final JavaSparkContext sparkContext;
+        private final PrestoSparkTaskCompilerFactory compilerFactory;
+        private final int hashPartitionCount;
+        private final CollectionAccumulator<byte[]> taskStatsCollector;
+        private final TableWriteInfo tableWriteInfo;
+
+        private RddFactory(
+                Session session,
+                JsonCodec<SparkTaskDescriptor> sparkTaskRequestJsonCodec,
+                JavaSparkContext sparkContext,
+                PrestoSparkTaskCompilerFactory taskCompilerFactory,
+                int hashPartitionCount,
+                CollectionAccumulator<byte[]> taskStatsCollector,
+                TableWriteInfo tableWriteInfo)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.sparkTaskRequestJsonCodec = requireNonNull(sparkTaskRequestJsonCodec, "sparkTaskRequestJsonCodec is null");
+            this.sparkContext = requireNonNull(sparkContext, "sparkContext is null");
+            this.compilerFactory = requireNonNull(taskCompilerFactory, "taskCompilerFactory is null");
+            this.hashPartitionCount = hashPartitionCount;
+            this.taskStatsCollector = requireNonNull(taskStatsCollector, "taskStatsCollector is null");
+            this.tableWriteInfo = requireNonNull(tableWriteInfo, "tableWriteInfo is null");
+        }
+
+        public JavaPairRDD<Integer, byte[]> createRdd(SubPlanWithTaskSources subPlan)
+        {
+            PlanFragment fragment;
+            if (subPlan.getFragment().getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
+                fragment = subPlan.getFragment().withBucketToPartition(Optional.of(IntStream.range(0, hashPartitionCount).toArray()));
+            }
+            else {
+                fragment = subPlan.getFragment();
+            }
+
+            checkArgument(!fragment.getStageExecutionDescriptor().isStageGroupedExecution(), "unexpected grouped execution fragment: %s", fragment.getId());
+
+            // scans
+            List<PlanNodeId> tableScans = fragment.getTableScanSchedulingOrder();
+
+            // source stages
+            List<RemoteSourceNode> remoteSources = fragment.getRemoteSourceNodes();
+            checkArgument(tableScans.isEmpty() || remoteSources.isEmpty(), "stage has both, remote sources and table scans");
+
+            if (!tableScans.isEmpty()) {
```

Comment:
I think the branches in this method should be similar to `createStageScheduler#createStageScheduler`, i.e. first branch on `partitioning` and see if its `SOURCE_DISTRIBUTION` :) 

We can do this refactor later given there are too many things, just put a TODO.

Reply:
Most likely the entire SparkRddFactory will be refactored once we try to support more query shapes. I wouldn't bother much right now trying to make it perfect.